### ISSUE TITLE
grrapi failing to parse items field

### ISSUE THREAD

**Issue by LunarMongoose** [google/grr#590]
I am a little new to grr and specifically I have been trying to get the api to work for me. I am using the github tutorial and I cannot get past an error I am getting.
I run the following commands
>>> grrapi = api.InitHttp(api_endpoint="http://localhost:4567", auth = ("admin","demo"))
>>> search_result = grrapi.SearchClients("")

and I get this stack trace and error

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/api.py", line 26, in SearchClients
    return client.SearchClients(query, context=self._context)
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/client.py", line 266, in SearchClients
    items = context.SendIteratorRequest("SearchClients", args)
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/context.py", line 48, in SendIteratorRequest
    first_page = pages.next()
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/context.py", line 31, in _GeneratePages
    result = self.connector.SendRequest(handler_name, args_copy)
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/connectors/http_connector.py", line 211, in SendRequest
    self._InitializeIfNeeded()
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/connectors/http_connector.py", line 104, in _InitializeIfNeeded
    self._FetchRoutingMap()
  File "/usr/local/lib/python2.7/dist-packages/grr_api_client/connectors/http_connector.py", line 79, in _FetchRoutingMap
    json_format.Parse(json_str, proto, ignore_unknown_fields=True)
  File "/usr/local/lib/python2.7/dist-packages/google/protobuf/json_format.py", line 369, in Parse
    return ParseDict(js, message, ignore_unknown_fields)
  File "/usr/local/lib/python2.7/dist-packages/google/protobuf/json_format.py", line 384, in ParseDict
    parser.ConvertMessage(js_dict, message)
  File "/usr/local/lib/python2.7/dist-packages/google/protobuf/json_format.py", line 415, in ConvertMessage
    self._ConvertFieldValuePair(value, message)
  File "/usr/local/lib/python2.7/dist-packages/google/protobuf/json_format.py", line 501, in _ConvertFieldValuePair
    raise ParseError('Failed to parse {0} field: {1}'.format(name, e))
google.protobuf.json_format.ParseError: Failed to parse items field: Failed to parse argsTypeDescriptor field: Failed to parse default field: Can not find message descriptor by type_url: type.googleapis.com/ApiGetGrrBinaryBlobArgs..

Any help is appreciated

**Comment by mbushkov** [comment]
Hi! Quick question - how did you install GRR server (from DEB, from PIP,
from source)? Which GRR server version do you have (it should be displayed
int the bottom right corner of AdminUI)? And the same question about
grr_api - how did you install it (from PIP, from source), which command did
you use?

Cheers,
Misha

**Comment by LunarMongoose** [comment]
I installed the server from the Docker image. It does register that it has a client and says the version is Grr 3.2.2.1. On the client I ran 
> pip install grr-api-client 
then I ran
> from grr_api_client import api
> grrapi = api.InitHttp(api_endpoint="http://localhost:4567",auth=("admin","demo"))
> search_result = grrapi.SearchClients("")
Then I got the big stack trace. I ran no other commands.
If it makes a difference both the client and the server are running in separate docker containers

**Comment by mbushkov** [comment]
I think there's a version mismatch here. If you used grr:latest Docker
image, then it has the HEAD version of GRR, while the released
grr-api-client PIP package corresponds to the stable GRR release version.
Can you please try installing grr-api-client from HEAD (i.e. from source)
as described here
<https://github.com/google/grr/tree/master/api_client/python#installation> and
try again?

Cheers,
Misha

**Comment by LunarMongoose** [comment]
WHen I try to install froom source on a docker container it runs into systemd errors. 




So I tried to install grr on my VM running ubuntu 16.04 and after gettting it registered with teh server and following the install from source instructions i get this error

lunar@lunar-VirtualBox:~$ source GRR_NEW/bin/activate
(GRR_NEW) lunar@lunar-VirtualBox:~$ cd grr/api_client/python
(GRR_NEW) lunar@lunar-VirtualBox:~/grr/api_client/python$ pip install --editable .
Obtaining file:///home/lunar/grr/api_client/python
Collecting grr_response_proto==3.2.2post1 (from grr-api-client==3.2.2.post1)
  Could not find a version that satisfies the requirement grr_response_proto==3.2.2post1 (from grr-api-client==3.2.2.post1) (from versions: 3.2.1.post4, 3.2.1.post7, 3.2.2.post0)
No matching distribution found for grr_response_proto==3.2.2post1 (from grr-api-client==3.2.2.post1)

**Comment by mbushkov** [comment]
Hm, I see.
What if you do the following before installing grr-api-client (i.e. before cd'ing to grr/api/client/python and running `pip install --editable .`)?

```
cd grr/grr/proto
pip install --editable .
```

This will install a local version of grr-response-proto. We recently made dependency  requirements for PIP packages stricter: versions of different GRR components have to match exactly. That's why you need to install grr-response-proto from HEAD as well (this is what the command above does).

**Comment by LunarMongoose** [comment]
I get this

>(GRR_NEW) lunar@lunar-VirtualBox:~/grr/grr/proto$ pip install --editable .
Obtaining file:///home/lunar/grr/grr/proto
    Complete output from command python setup.py egg_info:
    /usr/lib/python2.7/distutils/dist.py:267: UserWarning: Unknown distribution option: 'data'
      warnings.warn(msg)
    /home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/dist.py:398: UserWarning: Normalizing '3.2.2post1' to '3.2.2.post1'
      normalized_version,
    running egg_info
    creating grr_response_proto.egg-info
    writing requirements to grr_response_proto.egg-info/requires.txt
    writing grr_response_proto.egg-info/PKG-INFO
    writing top-level names to grr_response_proto.egg-info/top_level.txt
    writing dependency_links to grr_response_proto.egg-info/dependency_links.txt
    writing manifest file 'grr_response_proto.egg-info/SOURCES.txt'
    Traceback (most recent call last):
      File "makefile.py", line 91, in <module>
        MakeProto()
      File "makefile.py", line 71, in MakeProto
        "We need 3.3.0 not %s" % output)
    RuntimeError: Incompatible protoc compiler detected. We need 3.3.0 not libprotoc 3.0.0
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/lunar/grr/grr/proto/setup.py", line 91, in <module>
        setup(**setup_args)
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/__init__.py", line 129, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python2.7/distutils/core.py", line 151, in setup
        dist.run_commands()
      File "/usr/lib/python2.7/distutils/dist.py", line 953, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python2.7/distutils/dist.py", line 972, in run_command
        cmd_obj.run()
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/egg_info.py", line 278, in run
        self.find_sources()
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/egg_info.py", line 293, in find_sources
        mm.run()
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/egg_info.py", line 524, in run
        self.add_defaults()
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/egg_info.py", line 560, in add_defaults
        sdist.add_defaults(self)
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/py36compat.py", line 34, in add_defaults
        self._add_defaults_python()
      File "/home/lunar/GRR_NEW/local/lib/python2.7/site-packages/setuptools/command/sdist.py", line 128, in _add_defaults_python
        self.filelist.extend(build_py.get_source_files())
      File "/usr/lib/python2.7/distutils/command/build_py.py", line 302, in get_source_files
        return [module[-1] for module in self.find_all_modules()]
      File "/home/lunar/grr/grr/proto/setup.py", line 55, in find_all_modules
        compile_protos()
      File "/home/lunar/grr/grr/proto/setup.py", line 49, in compile_protos
        ["python", "makefile.py", "--clean"], cwd=THIS_DIRECTORY)
      File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['python', 'makefile.py', '--clean']' returned non-zero exit status 1
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /home/lunar/grr/grr/proto/

**Comment by mbushkov** [comment]
### Using latest released GRR docker version (grrdocker/grr:v3.2.2.0)

First, quick question: is it critical for you to use GRR HEAD version when using Docker? If you use the latest stable version (v3.2.2.0), grr-api-client package will work out of the box without problems.

For example, if you run:
```bash
sudo docker run \
  -e EXTERNAL_HOSTNAME="localhost" \
  -e ADMIN_PASSWORD="demo" --ulimit nofile=1048576:1048576 \
  -p 0.0.0.0:8000:8000 -p 0.0.0.0:8080:8080 \
  grrdocker/grr:v3.2.2.0 grr
```

Then the following should work without problems:
```bash
virtualenv .venv
source .venv/bin/activate
pip install grr-api-client
grr_api_shell --basic_auth_password=demo \
  --basic_auth_username=admin \
  http://localhost:8000
```

(you can test that the shell works by running this in IPython prompt):
```python
list(grrapi.SearchClients())
```

### To use grr:latest

If for some reason you really need to run GRR from HEAD, you'll have to build grr-response-proto and grr-api-client PIP packages from source yourself. To do that:

```bash
git clone https://github.com/google/grr
cd grr
virtualenv .venv
source .venv/bin/activate
# This will put protoc 3.3.0 into $HOME/protobuf folder.
bash -x ./travis/install_protobuf.sh linux
# This installs grr-response-proto package.
env PROTOC=${HOME}/protobuf/bin/protoc pip install -e grr/proto
# This installs grr-api-client package
pip install -e api_client/python/
# This runs grr_api_shell binary so that you can test that everything works.
grr_api_shell --basic_auth_password=demo --basic_auth_username=admin \
  http://localhost:8000
```

Hope this helps!

I've tested that both suggested approaches work on Ubuntu 16. I'll update GRR docs soon to indicate the differences between release and HEAD versions and to describe possible corresponding issues.

**Comment by LunarMongoose** [comment]
Thank you! that seemed to work for me. It was not critical to use the latest version of server. It just seems that I got the latest version. It really was just a version mismatch between client api and released server. Hope this helps someone else too
